doc/sphinx/proof-engine/ltac.rst

```diff
@@ -451,178 +461,185 @@ structure:
    ``v`` is applied but only its first success is used. If ``v`` fails,
-   :n:`once @ltac_expr` fails like ``v``. If ``v`` has at least one success,
-   :n:`once @ltac_expr` succeeds once, but cannot produce more successes.
+   :tacn:`once` :n:`@ltac_expr3` fails like ``v``. If ``v`` has at least one success,
+   :tacn:`once` :n:`@ltac_expr3` succeeds once, but cannot produce more successes.
 
-Checking the successes
-~~~~~~~~~~~~~~~~~~~~~~
+   :tacn:`once` is an :token:`ltac_expr3`.
+
+Checking for a single success
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 
 Coq provides an experimental way to check that a tactic has *exactly
 one* success:
 
-.. tacn:: exactly_once @ltac_expr
+.. tacn:: exactly_once @ltac_expr3
    :name: exactly_once
 
-   :n:`@ltac_expr` is evaluated to ``v`` which must be a tactic value. The tactic value
+   :n:`@ltac_expr3` is evaluated to ``v`` which must be a tactic value. The tactic value
    ``v`` is applied if it has at most one success. If ``v`` fails,
-   :n:`exactly_once @ltac_expr` fails like ``v``. If ``v`` has a exactly one success,
-   :n:`exactly_once @ltac_expr` succeeds like ``v``. If ``v`` has two or more
-   successes, :n:`exactly_once @ltac_expr` fails.
+   :tacn:`exactly_once` :n:`@ltac_expr3` fails like ``v``. If ``v`` has a exactly one success,
+   :tacn:`exactly_once` :n:`@ltac_expr3` succeeds like ``v``. If ``v`` has two or more
+   successes, :tacn:`exactly_once` :n:`@ltac_expr3` fails.
+
+   :tacn:`exactly_once` is an :token:`ltac_expr3`.
 
    .. warning::
 
       The experimental status of this tactic pertains to the fact if ``v``
-      performs side effects, they may occur in an unpredictable way. Indeed,
+      has side effects, they may occur in an unpredictable way. Indeed,
       normally ``v`` would only be executed up to the first success until
-      backtracking is needed, however exactly_once needs to look ahead to see
+      backtracking is needed, however :tacn:`exactly_once` needs to look ahead to see
       whether a second success exists, and may run further effects
       immediately.
 
    .. exn:: This tactic has more than one success.
       :undocumented:
 
-Checking the failure
+Checking for failure
 ~~~~~~~~~~~~~~~~~~~~
 
 Coq provides a derived tactic to check that a tactic *fails*:
 
-.. tacn:: assert_fails @ltac_expr
+.. tacn:: assert_fails @ltac_expr3
    :name: assert_fails
 
-   This behaves like :tacn:`idtac` if :n:`@ltac_expr` fails, and
-   behaves like :n:`fail 0 @ltac_expr "succeeds"` if :n:`@ltac_expr`
+   Behaves like :tacn:`idtac` if :n:`@ltac_expr3` fails, and
+   behaves like :tacn:`fail` :n:`0 @ltac_expr3 "succeeds"` if :n:`@ltac_expr`
    has at least one success.
 
-Checking the success
+   .. todo PR the examples for assert_fails and assert_succeeds seem incorrect;
+      fail can only take an ident rather than an ltac_expr3.  Also shouldn't have
+      a printable string; IIUC that's not equivalent to assert_* behavior.
```

Comment:
Was probably meant to mean that it prints the tactic name as part of the error message (as a string). But I don't see why it is so important to explain precisely what the error message is.

Reply:
Will leave it for later

Reply:
Indeed, it's not so important to explain what the error message is, and it's fine to just say that it fails when the argument has at least one success.

Reply:
(And I think my hope was that it would eventually print the tactic name as part of the error message, but right now I think it ends up just printing the literal string "<tactic expression> succeeds" (or something like that).